src/ipc/ipc4/helper.c

```diff
@@ -296,33 +296,34 @@ static int ipc_pipeline_module_free(uint32_t pipeline_id)
 
 	icd = ipc_get_comp_by_ppl_id(ipc, COMP_TYPE_COMPONENT, pipeline_id, IPC_COMP_ALL);
 	while (icd) {
-		struct list_item *list, *_list;
 		struct comp_buffer *buffer;
+		struct comp_buffer *next_buffer;
 
 		/* free sink buffer allocated by current component in bind function */
-		list_for_item_safe(list, _list, &icd->cd->bsink_list) {
-			struct comp_dev *sink;
-
-			buffer = container_of(list, struct comp_buffer, source_list);
-			pipeline_disconnect(icd->cd, buffer, PPL_CONN_DIR_COMP_TO_BUFFER);
-			sink = buffer->sink;
+		buffer = comp_dev_get_first_data_consumer(icd->cd);
+		while (buffer) {
+			next_buffer = comp_dev_get_next_data_consumer(icd->cd, buffer);
+			struct comp_dev *sink = comp_buffer_get_sink_component(buffer);
 
 			/* free the buffer only when the sink module has also been disconnected */
 			if (!sink)
 				buffer_free(buffer);
+
+			buffer = next_buffer;
 		}
 
 		/* free source buffer allocated by current component in bind function */
-		list_for_item_safe(list, _list, &icd->cd->bsource_list) {
-			struct comp_dev *source;
-
-			buffer = container_of(list, struct comp_buffer, sink_list);
+		buffer = comp_dev_get_first_data_producer(icd->cd);
+		while (buffer) {
```

Comment:
cannot use your `comp_buf_for_each...` here? Maybe at least `for (buffer = comp_dev_get_first_data_producer(icd->cd); buffer; buffer = next_buffer)`

Reply:
comp_buf_for_each is "list_for_item", not "list_for_item_safe" and will fail on "use-after-free"
you think such FOR is more readable than while?